test.rb

```diff
@@ -0,0 +1,211 @@
+# frozen_string_literal: true
+
+require 'zlib'
+require 'stackprof'
+require 'googleauth'
+require 'google/apis/cloudprofiler_v2'
+
+require './profile_pb'
+
+Cloudprofiler = Google::Apis::CloudprofilerV2
+
+PROFILE_TYPES = {
+  'CPU' => :cpu,
+  'WALL' => :wall,
+  'HEAP_ALLOC' => :object
+}.freeze
+SERVICE_REGEXP = /^[a-z]([-a-z0-9_.]{0,253}[a-z0-9])?$/.freeze
+
+SCOPES = ['https://www.googleapis.com/auth/cloud-platform'].freeze
+
+def parse_duration(d)
+  m = /^(\d+)s$/.match(d)
+  return 10 if m.nil?
+
+  Integer(m[1])
+end
+
+class Profiler
+  def initialize(service:, project_id:, serviceVersion: nil, debug_logging: false, instance: nil, zone: nil)
+    raise "service must match #{SERVICE_REGEXP}" unless SERVICE_REGEXP =~ service
+
+    @service = service
+    @project_id = project_id
+    @debug_logging = debug_logging
+
+    @labels = { language: 'ruby' }
+    @labels[:version] = serviceVersion unless serviceVersion.nil?
+    @labels[:zone] = zone unless zone.nil?
+
+    @deployment = Cloudprofiler::Deployment.new(project_id: project_id, target: service, labels: @labels)
+
+    @profile_labels = {}
+    @profile_labels[:instance] = instance unless instance.nil?
+
+    @profiler = Cloudprofiler::CloudProfilerService.new
+    @profiler.authorization = Google::Auth.get_application_default(SCOPES)
+  end
+
+  attr_reader :service, :project_id, :labels, :deployment, :profile_labels
+
+  def debug_log(message)
+    puts(message) if @debug_logging
+  end
+
+  def create_profile
+    req = Cloudprofiler::CreateProfileRequest.new(deployment: deployment, profile_type: PROFILE_TYPES.keys)
+    debug_log('creating profile')
+    @profiler.create_profile("projects/#{deployment.project_id}", req)
+  end
+
+  def update_profile(profile)
+    debug_log('updating profile')
+    @profiler.patch_project_profile(profile.name, profile)
+    debug_log('profile updated')
+  end
+
+  def profile_and_upload(profile)
+    debug_log("profiling #{profile.profile_type} for #{profile.duration}")
+    profile.profile_bytes = profile(parse_duration(profile.duration), PROFILE_TYPES[profile.profile_type])
```

Comment:
If you expect `PROFILE_TYPES[profile.profile_type]` to find something, then use `.fetch` as it will raise a KeyError if it doesn't find something. If it can return `nil`, then its better to use `.fetch` with a default second argument of `nil` (e.g. `.fetch(profile.profile_type, nil)`) so that it's explicit the value can sometimes be `nil`. Generally you should always use `.fetch` when accessing hash values.
 
```suggestion
    profile.profile_bytes = profile(parse_duration(profile.duration), PROFILE_TYPES.fetch(profile.profile_type))
```

There's quite a few places lower down that you should use `fetch` as well.